memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -109,55 +109,56 @@ func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Res
 		return reconcile.Result{}, err
 	}
 
-	// Check if the Deployment already exists, if not create a new one
-	deployment := &appsv1.Deployment{}
-	err = r.client.Get(context.TODO(), types.NamespacedName{Name: memcached.Name, Namespace: memcached.Namespace}, deployment)
-	if err != nil && errors.IsNotFound(err) {
-		// Define a new Deployment
-		dep := r.deploymentForMemcached(memcached)
-		reqLogger.Info("Creating a new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-		err = r.client.Create(context.TODO(), dep)
-		if err != nil {
-			reqLogger.Error(err, "Failed to create new Deployment.", "Deployment.Namespace", dep.Namespace, "Deployment.Name", dep.Name)
-			return reconcile.Result{}, err
-		}
-		// Deployment created successfully - return and requeue
-		// NOTE: that the requeue is made with the purpose to provide the deployment object for the next step to ensure the deployment size is the same as the spec.
-		// Also, you could GET the deployment object again instead of requeue if you wish. See more over it here: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile#Reconciler
-		return reconcile.Result{Requeue: true}, nil
-	} else if err != nil {
-		reqLogger.Error(err, "Failed to get Deployment.")
-		return reconcile.Result{}, err
-	}
+	// Update Deployment if it has changed to something other than expected. If the Deployment
+	// doesn't exist it will be created (initial creation is not really a special case).
+	deployment := &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: memcached.Name, Namespace: memcached.Namespace}}
+	op, err := controllerutil.CreateOrUpdate(context.TODO(), r.client, deployment, func(ignore runtime.Object) error {
+
+		// Set deployment object to expected state in memory. If it's different than what
+		// the calling `CreateOrUpdate` function gets from Kubernetes, it will send an
+		// update request back to the apiserver with the expected state determined here.
+		deploymentForMemcached(memcached, deployment)
```

Comment:
The if to check if is new or nor could be outside, for example, and its name could be changed to make clear that it will just create the Deployment resource for the Memcached.
 
Then, if is not new we could call the alidateMemcachedSpecs, for example, which would return the object updated or an error. WDYT? Would not it make more clear what is occurring here?

Reply:
@camilamacedo86 I might be misunderstanding your suggestion, but it sounds like you're suggesting that we have separate functions and checks to handle create and update separately?

If so, I disagree with that approach. The point of the `CreateOrUpdate` function is that there is no difference between the Create and Update cases in terms of reconciling this CR, so I think this is well-designed and idiomatic as is.

In fact, the `Reconcile()` interface was intentionally designed to make it difficult to determine what event caused the reconciliation, because the event that caused it doesn't matter. The goal of `Reconcile()` is to read the current state of the cluster and to ensure that it matches the desired state.

Reply:
I think is easier to do the suggestions than explain sometimes :-) So, let's do one thing @joelanford if/or when this PR be ok for you is ook for me and after we merge it I can do some PR with my suggestions and etc.. WDYT? 

PS.: the getting started probably will get outdated with these changes as well.